### DISCUSSION TITLE
Changesets: Schema versioning and type overriding

### DISCUSSION THREAD

**Discussion by caspervk** [strawberry-graphql/strawberry#3619] [Q&A]
I would like to be able to make breaking changes to my GraphQL API. I am doing this by creating different schemas and exposing them at `/graphql/v1`, `/graphql/v2` etc.

Most changes are relatively minor and implemented by subclassing the latest schema and overwriting a single mutator or changing a field. Earlier versions are implemented by subclassing the next version, leading to a chain of schemas all the way to the latest one. The latest schema is the "main" one, containing most of the logic, so that earlier versions can be deprecated and deleted. This isn't pretty, but it works. The code is open source, for example: https://github.com/magenta-aps/os2mo/blob/1f557a658145200e090c4805eb8fe24df6761bc1/backend/mora/graphapi/versions/v21/version.py.

Sometimes, subclassing the schema doesn't cut it. For example, if I need to change a field on a type which is used both directly in the top-level query as well as nested as fields on other types. In this case, I can overwrite the field and type on the top-level query easily, but I also have to overwrite in any fields which reference it, and then any fields which reference _those_ models. The number of required changes quickly explodes.

Is there a better way to do this? Does the schema have a store of all types which I could use to overwrite one type with another everywhere in the schema at once?

---

**EDIT**: Subclassing is a bad idea. Please see [my reply](https://github.com/strawberry-graphql/strawberry/discussions/3619#discussioncomment-10728114) about GraphQL Ruby's [Changesets](https://graphql-ruby.org/changesets/overview.html).

**Comment by DoctorJohn** [comment]
From my experience, it's unusual in GraphQL to have the whole endpoint versioned. Instead, I recommend continuously updating a single schema. This is also mentioned on the [GraphQL website](https://graphql.org/learn/best-practices/#versioning):

> common practice of always avoiding breaking changes and serving a versionless API.

Naturally, adding new types or fields is a non-breaking change in GraphQL. To deprecate old fields/types, you can use the `@deprecated` directive. If you have to change your schema in a breaking way, you could start versioning on a type or field level.

Have you considered this approach, or is versioning the endpoint the only option in your case?

**Reply by caspervk** [reply]
Our API exposes a bi-temporal data model which is already difficult for consumers to understand - but even more importantly, which is difficult to properly design. I'm worried that introducing additional types or fields will bloat the schema and make it even harder to understand while we iteratively figure out the best design.

I feel like the recommendation on the GraphQL website assumes that you know how to design the perfect schema from the beginning. I'm not a 7-figure 10x silicon valley facebook engineer, and so is not smart enough to do so. For example, we needed to introduce pagination on every field and ended up duplicating the entire schema and accompanying strawberry code. Doing this in a non-breaking way would involve duplicating nearly every type and field in the schema under a new name. Since we initially chose the best-fitting name for the types and fields, we would be forced to choose worse and worse names every time we realise the need for a breaking change. This quickly results in a very hard-to-understand and messy schema.

My ideal solution is to hide our past mistakes in older versions and only expose the best possible version to (new) customers. It seems to me some of the cool silicon valley engineers have realised this as well:
- [Shopify](https://shopify.dev/graphiql/admin-graphiql) has explicit versioning of their schema exactly like we implement today.
- [GitHub](https://docs.github.com/en/graphql/overview/breaking-changes) does breaking changes to their GraphQL schema all the time.

**Reply by erikwrede** [reply]
I understand your desire to reduce complexity in the schema now by versioning, but usually that is just a tradeoff for more complexity later on. Let me explain.

Contrary to typical RESTful endpoints, a type might be re-used all over the graph, being queriable from different types and mutations. This makes it harder to keep track of usage and compatibility throughout your schema. While versioning seems to be a simple fix, you should keep in mind that all of the past versions still use a portion of this type - maybe with different names, maybe with incompatible changes. Keeping track of this over multiple schemas is **hard**. I strongly suggest not taking on frequent versioning unless there is enough GraphQL experience in your company to handle the implications and complexity. Depending on the amount of clients and size of the app, versioning is simply a lot of added complexity with little value. Meta, who invented GraphQL, also doesn't do endpoint versioning and they've been running fine for years. Similar to many other enterprises. Shopify on the other hand makes their API available for Customer use with strict deprecation policies (no more than 4 Versions live at the same time). This _may_ require more api stability that is achieved by versioning than internal use of the GraphQL endpoint.

Instead of versioning, I'd recommend to
- add new fields `fieldNameV2` when adding pagination
- Add new V2 fields when making breaking changes to a field on a type
- only add new V2 types if the new type is completely incompatible to the old type
- have a deprecation policy per field
- publish a schema excluding deprecated fields to your consumers (you can filter the schema by the `@deprecated` directive)
- have usage monitoring to see which clients may be broken by the deprecation

The V2 fields also make it super easy for your clients to migrate over to the new versions step-by-step without switching over to an entirely new API

With regard to versioning, any further tooling to make that easier by schema inheritance is currently out of scope for us.

**Reply by caspervk** [reply]
Thank you for taking the time to write such a thorough reply!

I found this talk by Shopify which explains their versioning scheme: https://gitnation.com/contents/hard-graphql-problems-at-shopify. Of particular interest is the _Handling Deprecated Fields and Nullability_  and _Nullability and Shopify's Versioning System_ sections:

> Around the V2 and V3 naming things, that was because of a limitation that we had in our versioning of process, that is now fixed. So hopefully those will never happen again, because frankly they're ugly and they're very confusing. Like which do I use? Price V2, V3, but I'm using the January release of the schema, but I'm using V2, it's just a mess, right? We all love to have nice APIs that are named nicely and are very OCD about things being consistent. So dropping V2s and V3s are something we definitely wanna do.

> you have to follow through on the process of like make a v2 field, get all your clients onto that v2 field, then get rid of your old field or change your old field and then migrate them back to the original field, right? So I think it's like always a long process to get rid of those v2 v3s in your code base.

> basically we're making our versioning system that much more powerful to allow us to make these breaking changes in a nicer, more fluid way. That author v2 would not have to be a thing. It would literally just, if you request version A, it's not a null, you request version B in the future, it's suddenly null. 

> We're a little bit lucky because of the GraphQL Ruby Gen that we have supports a lot of things that we utilize to do versioning.

The feature he is referring to here is GraphQL Ruby's [Changesets](https://graphql-ruby.org/changesets/overview.html):

> Out-of-the-box, GraphQL is [versionless by design](https://graphql.org/learn/best-practices/#versioning). GraphQL’s openness to extension paves the way for continuously expanding and improving an API. You can always add new fields, new arguments, and new types to implement new features and customize existing behavior.
>
> However, sometimes a business case may call for a different versioning scheme. [GraphQL-Enterprise](https://graphql.pro/enterprise)’s “Changesets” enable schemas to release any change – even breaking changes – to clients, depending on what version of the schema they’re using. With Changesets, you can redefine existing fields, define new types using old names, add or remove enum values – anything, really – while maintaining compatibility for existing clients.
>
> Changesets are a complementary evolution technique to continuous additions. **In general, additive changes (new fields, new arguments, new types) are best added right to the existing schema. But if you need to remove something from the schema or redefine existing parts of the schema in non-backwards-compatible ways, Changesets provide a handy way of doing so.**

From https://graphql-ruby.org/changesets/definition.html:

> New things can be published in a changeset by adding `added_in: SomeChangeset` to their configuration. For example, to add a new argument to a field:
> ```ruby
> field :search_recipes, [Types::Recipe] do
>   argument :query, String
>   argument :tags, [Types::RecipeTag], required: false, added_in: Changesets::AddRecipeTags
> end
> ```
>
> You can also provide a replacement implementation by using `added_in:`. When a new definition has the same name as an existing definition, it implicitly replaces the previous definition in new versions of the API. For example:
> ```ruby
> field :rating, Integer, "A 1-5 score for this recipe" # This definition will be superseded by the following one
> field :rating, Float, "A 1.0-5.0 score for this recipe", added_in: Changesets::FloatingPointRatings
> ```
> Here, a new implementation for `rating` will be used when clients requests an API version that includes `Changesets::FloatingPointRatings`. (If the client requests a version before that changeset, then the preceding implementation would be used instead.)

A lot of other interesting examples are available at https://graphql-ruby.org/changesets/definition.html#examples:

> [Fields](https://graphql-ruby.org/changesets/definition.html#fields): adding, modifying, and removing fields
>     [Arguments](https://graphql-ruby.org/changesets/definition.html#arguments): adding, modifying, and removing arguments
>     [Enum values](https://graphql-ruby.org/changesets/definition.html#enum-values): adding, modifying, and removing arguments
>     [Unions](https://graphql-ruby.org/changesets/definition.html#unions): adding or removing object types from a union
>     [Interfaces](https://graphql-ruby.org/changesets/definition.html#interfaces): adding or removing interface implementations from object types
>     [Types](https://graphql-ruby.org/changesets/definition.html#types): changing one type definition for another
>     [Runtime](https://graphql-ruby.org/changesets/definition.html#runtime): choosing a behavior at runtime based on the current request and changeset

---

I understand that schema versioning by inheritance is currently out of scope for the Strawberry project, but is there a way I could implement something similar to Ruby GraphQL's changesets in my Strawberry application?

**Reply by caspervk** [reply]
@erikwrede would a versioning scheme based on changesets also be out of scope of the project, or should I perhaps open a feature request?

**Reply by jirikuncar** [reply]
The following works for our use case. 
```py
@strawberry.type
class Query:
   v1: QueryV1
   v2: QueryV2
```
You still get a huge schema but you can reuse types that haven't changed between versions.

**Reply by caspervk** [reply]
Ideally I would like to avoid duplication and reuse as much code as possible between the versions. I'm not sure if the scheme you are suggesting enables this. Like in my original post, changing just one field can quickly explode when you have to overwrite it in any field which reference it, and then any field which reference _those_ fields and so on.

I feel like GraphQL Ruby's [Changesets](https://graphql-ruby.org/changesets/overview.html) was a true revelation -- I just don't know how to implement it in Strawberry.

**Reply by Skeen** [reply]
@jirikuncar The approach you're suggesting also does _not_ work for mutations, which per [Section 6.2.2 og the October 2021 GraphQL spec](https://spec.graphql.org/October2021/#sec-Mutation) should be executed serially. Nesting mutations under a top-level version type breaks this seriality, and thus breaks mutation behavior.

If https://github.com/graphql/graphql-spec/issues/252 was to be merged (unlikely seeing at it is ~8 years old at this point), this would change the behavior and make your suggested approach feasible for mutations, but as it stands now, the GraphQL spec does not support it.
